persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/StargateQueryHandler.java

```diff
@@ -64,15 +87,188 @@ public ResultMessage process(
+          castStatement.keyspace(),
+          castStatement.columnFamily());
+
+      try {
+        authorization.authorizeSchemaWrite(
+            authToken, castStatement.keyspace(), castStatement.columnFamily(), Scope.TRUNCATE);
+      } catch (io.stargate.auth.UnauthorizedException e) {
+        throw new UnauthorizedException(
+            String.format(
+                "No TRUNCATE permission on <table %s.%s>",
+                castStatement.keyspace(), castStatement.columnFamily()));
+      }
+
+      logger.debug(
+          "authorized statement of type {} on {}.{}",
+          castStatement.getClass().toString(),
+          castStatement.keyspace(),
+          castStatement.columnFamily());
+    } else if (statement instanceof SchemaAlteringStatement) {
+      SchemaAlteringStatement castStatement = (SchemaAlteringStatement) statement;
+      Scope scope = null;
+      String keyspaceName = null;
+      String tableName = null;
+
+      if (statement instanceof CreateTableStatement) {
+        scope = Scope.CREATE;
+        keyspaceName = castStatement.keyspace();
+        tableName = castStatement.columnFamily();
+      } else if (statement instanceof DropTableStatement) {
+        scope = Scope.DELETE;
+        keyspaceName = castStatement.keyspace();
+        tableName = castStatement.columnFamily();
+      } else if (statement instanceof AlterTableStatement) {
+        scope = Scope.ALTER;
+        keyspaceName = castStatement.keyspace();
+        tableName = castStatement.columnFamily();
+      } else if (statement instanceof CreateKeyspaceStatement) {
+        scope = Scope.CREATE;
+        keyspaceName = castStatement.keyspace();
+        tableName = null;
+      } else if (statement instanceof DropKeyspaceStatement) {
+        scope = Scope.DELETE;
+        keyspaceName = castStatement.keyspace();
+        tableName = null;
+      } else if (statement instanceof AlterKeyspaceStatement) {
+        scope = Scope.ALTER;
+        keyspaceName = castStatement.keyspace();
+        tableName = null;
+      }
+
+      logger.debug(
+          "preparing to authorize statement of type {} on {}.{}",
+          castStatement.getClass().toString(),
+          keyspaceName,
+          tableName);
+
+      try {
+        authorization.authorizeSchemaWrite(authToken, keyspaceName, tableName, scope);
+      } catch (io.stargate.auth.UnauthorizedException e) {
+        throw new UnauthorizedException(
+            String.format(
+                "Missing correct permission on %s.%s",
+                keyspaceName, (tableName == null ? "" : tableName)));
+      }
+
+      logger.debug(
+          "authorized statement of type {} on {}.{}",
+          castStatement.getClass().toString(),
+          keyspaceName,
+          tableName);
+    }
+  }
+
+  private AuthorizationService getAuthorizationServiceFromContext() {
+    BundleContext bundleContext = FrameworkUtil.getBundle(this.getClass()).getBundleContext();
```

Comment:
Adding OSGi dependencies this deep in a persistence module code looks a bit awkward to me... Is there a way to set the auth service reference at the handler creation time? I suppose it should be available when the handler is created.

At creation site we might be able to pass the auth reference through a thread-local variable, thus removing OSGi dependencies from this class. WDYT?

Reply:
I know, I felt dirty when I did it. The problem is if we try to register the service in the io.stargate.db.cassandra.CassandraPersistenceActivator it'll create a circular dependency for auth-table-based-service.

Reply:
We could pass around an `AtomicReference<AuthorizationService>` at init time and set it (from the activator) when the service is available, perhaps?

Reply:
Done